segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadIndexCache.java

```diff
@@ -0,0 +1,382 @@
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Truncate the read index for given segment by removing all the chunks that are below given offset.
+     *
+     * @param streamSegmentName
+     * @param startOffset
+     */
+    public void truncateReadIndex(String streamSegmentName, long startOffset) {
+        val segmentReadIndex = getSegmentReadIndex(streamSegmentName);
+        if (null != segmentReadIndex) {
+            if (segmentReadIndex.chunkIndex.size() > 0) {
+                val headMap = segmentReadIndex.chunkIndex.headMap(startOffset);
+                if (null != headMap) {
+                    int removed = 0;
+                    ArrayList<Long> keysToRemove = new ArrayList<Long>();
+                    keysToRemove.addAll(headMap.keySet());
+                    for (val keyToRemove : keysToRemove) {
+                        segmentReadIndex.chunkIndex.remove(keyToRemove);
+                        removed++;
+                    }
+                    if (removed > 0) {
+                        totalChunkCount.getAndAdd(-1 * removed);
+                    }
+                }
+            }
+            if (segmentReadIndex.chunkIndex.size() > 0) {
+                segmentReadIndex.setGeneration(currentGeneration.get());
+            }
+        }
+    }
+
+    /**
+     * Evicts segments from the cache.
+     */
+    private void evictSegmentsFromOldestGeneration() {
+        val oldGen = oldestGeneration.get();
+        currentGeneration.getAndIncrement();
+
+        val iterator = segmentsToReadIndexMap.entrySet().iterator();
+        int total = totalChunkCount.get();
+        int removed = 0;
+        while (iterator.hasNext() && (segmentsToReadIndexMap.size() >= maxIndexedSegments || total >= maxIndexedChunks)) {
+            val entry = iterator.next();
+            if (entry.getValue().getGeneration() <= oldGen) {
+                val size = entry.getValue().chunkIndex.size();
+                removed += size;
+                total -= size;
+                iterator.remove();
+            }
+        }
+        if (removed > 0) {
+            oldestGeneration.compareAndSet(oldGen, oldGen + 1);
+            totalChunkCount.getAndAdd(-1 * removed);
+        }
+    }
+
+    /**
+     * Evicts chunks from the cache.
+     */
+    private void evictChunks(String streamSegmentName, long toRemoveCount) {
+        val segmentReadIndex = segmentsToReadIndexMap.get(streamSegmentName);
+        evictChunks(segmentReadIndex, toRemoveCount);
+    }
+
+    /**
+     * Evicts chunks from the cache.
+     */
+    private void evictChunks(SegmentReadIndex segmentReadIndex, long toRemoveCount) {
+        // Increment generation.
+        val previousGen = currentGeneration.getAndIncrement();
+        val oldGen = oldestGeneration.get();
+
+        // Step 1 : Go through all entries once to record counts per each generation.
```

Comment:
What's the cost of iterating through all entries upon eviction?

Reply:
This loop needs improvement. Let me think about this.

Reply:
Although chunks are removed on per segment basis.

Reply:
I'll re-implement the readIndex in next iteration. 
Currently the operation is O(number of chunks in segment that are cached). 
This is bounded by the max number of  cached index entries per segment 1024 . 
So far in my testing I have seen very low number of actual chunks per segments for NFS.
Immutable storage like S3 will have higher number of chunks per segment and this index needs to improve before that.   I'll create a issue to track this.

Reply:
I'm unresolving this thread as it doesn't look like it has a resolution. I mentioned this thread in the first comment of this file.

Reply:
I was planning to re-implement this cache in a separate PR after this PR.

Reply:
Can you create an issue and post the link here, please?

Reply:
https://github.com/pravega/pravega/issues/4902